htlcswitch/quiescer_test.go

```diff
@@ -0,0 +1,472 @@
+}
+
+// TestQuiescenceCantSendUpdatesAfterStfu tests that we can send channel updates
+// prior to but not after we send Stfu.
+func TestQuiescenceCantSendUpdatesAfterStfu(t *testing.T) {
+	t.Parallel()
+
+	harness := initQuiescerTestHarness(lntypes.Local)
+	require.True(t, harness.quiescer.CanSendUpdates())
+
+	msg := lnwire.Stfu{
+		ChanID:    cid,
+		Initiator: true,
+	}
+
+	err := harness.quiescer.RecvStfu(msg, harness.pendingUpdates.Remote)
+	require.NoError(t, err)
+
+	err = harness.quiescer.SendOwedStfu(harness.pendingUpdates.Local)
+	require.NoError(t, err)
+
+	require.False(t, harness.quiescer.CanSendUpdates())
+}
+
+// TestQuiescenceStfuNotNeededAfterRecv tests that after we receive an Stfu we
+// do not needStfu either before or after receiving it if we do not initiate
+// quiescence.
+func TestQuiescenceStfuNotNeededAfterRecv(t *testing.T) {
+	t.Parallel()
+
+	harness := initQuiescerTestHarness(lntypes.Local)
+
+	msg := lnwire.Stfu{
+		ChanID:    cid,
+		Initiator: true,
+	}
+	require.False(t, harness.quiescer.NeedStfu())
+
+	require.NoError(
+		t, harness.quiescer.RecvStfu(
+			msg, harness.pendingUpdates.Remote,
+		),
+	)
+
+	require.False(t, harness.quiescer.NeedStfu())
+}
+
+// TestQuiescenceInappropriateMakeStfuReturnsErr ensures that we cannot call
+// makeStfu at times when it would be a protocol violation to send it.
+func TestQuiescenceInappropriateMakeStfuReturnsErr(t *testing.T) {
+	t.Parallel()
+
+	harness := initQuiescerTestHarness(lntypes.Local)
+
+	harness.pendingUpdates.SetForParty(lntypes.Local, 1)
+
+	require.True(
+		t, harness.quiescer.MakeStfu(
+			harness.pendingUpdates.Local,
+		).IsErr(),
+	)
+
+	harness.pendingUpdates.SetForParty(lntypes.Local, 0)
+	msg := lnwire.Stfu{
+		ChanID:    cid,
+		Initiator: true,
+	}
+	require.NoError(
+		t, harness.quiescer.RecvStfu(
+			msg, harness.pendingUpdates.Remote,
+		),
+	)
+	require.True(
+		t, harness.quiescer.MakeStfu(
+			harness.pendingUpdates.Local,
+		).IsOk(),
+	)
+
+	require.NoError(
+		t, harness.quiescer.SendOwedStfu(harness.pendingUpdates.Local),
+	)
+	require.True(
+		t, harness.quiescer.MakeStfu(
+			harness.pendingUpdates.Local,
+		).IsErr(),
+	)
+}
+
+// TestQuiescerTieBreaker ensures that if both parties attempt to claim the
+// initiator role that the result of the negotiation breaks the tie using the
+// channel initiator.
+func TestQuiescerTieBreaker(t *testing.T) {
+	t.Parallel()
+
+	for _, initiator := range []lntypes.ChannelParty{
+		lntypes.Local, lntypes.Remote,
+	} {
+		harness := initQuiescerTestHarness(initiator)
+
+		msg := lnwire.Stfu{
+			ChanID:    cid,
+			Initiator: true,
+		}
+
+		req, res := fn.NewReq[fn.Unit, fn.Result[lntypes.ChannelParty]](
+			fn.Unit{},
+		)
+
+		harness.quiescer.InitStfu(req)
+		require.NoError(
+			t, harness.quiescer.RecvStfu(
+				msg, harness.pendingUpdates.Remote,
+			),
+		)
+		require.NoError(
+			t, harness.quiescer.SendOwedStfu(
+				harness.pendingUpdates.Local,
+			),
+		)
+
+		party := <-res
```

Comment:
I rarely use `<-` without a select and a timeout, just to make sure in the worst case scenario we can get back a testing result quickly.